### ISSUE TITLE
ESS evolution refactor

### ISSUE THREAD

**Issue by ndmlny-qs** [arviz-devs/arviz#1969]
## Tell us about it

In `plots/essplot.py` the ESS evolution calculation is performed in the `plot_ess`
method. I would like to move the calculation to `stats/diagnostics.py` where other ESS
calculations are done. Making this change would separate out diagnostic calculations
from the plotting implementation, and maintain a defined separation between what are
calculations, and plotting by ArviZ.

This is a simple change, however, I discuss below a possible restructuring of the
library, which is a larger change, but falls within the scope of separating calculations
from plotting.

## Thoughts on implementation

There are a lot of `_ess_*` methods in the `diagnostics.py` module in the `stats`
sub-package. I would like to open the conversation about possibly moving the
`diagnostics.py` module out of the `stats` package and into its own sub-package at the
same level as `stats`. The proposed new structure would look like the following;

```
arviz
├── data
├── diagnostics
│   ├── __init__.py
│   ├── ess.py
│   └── ...
├── plots
├── static
├── stats
├── tests
├── wrappers
├── __init__.py
├── labels.py
├── rcparams.py
├── sel_utils.py
└── utils.py
```

where the new sub-package `diagnostics` would break up diagnostic calculations into
their own modules. ESS is a primary suggestion, however, other items in the current
structure could get their own module. `r_hat` is another example that contains many
methods in the `diagnostics.py` module that would be suited to have its own module.

### Goals for the restructuring

- Maintain a defined separation of tasks; plotting _vs_ calculating.
- Allow users to run diagnostic calculations using ArviZ, and then use the results in
  custom plots.
- Separating calculations from plotting also allows users to create diagnostics plots
  with different plotting libraries other than the ones ArviZ currently supports.

@OriolAbril and @aloctavodia this is the first issue I'd like to propose from our
discussion about making bokeh widgets for model diagnostics. The larger refactor would
more than likely cause breaking changes to some user's code. This is not ideal in
general, but I can attempt to ensure backwards compatibility if the larger refactor is
approved.

**Comment by OriolAbril** [comment]
I agree with moving computations out of the plotting function. 

Not sure about what would be gained with the extra restructuring. Already the separation between diagnostics, stats and stats_utils (files) feels quite artificial. More prominently, summary combines both "stats" and "diagnostics", but also autocorr could easily be seen as a diagnostic, psislw as a stats_util, or even different methods of ess could be categorized differently (i.e. bulk as diagnostic, mean as stats). 

Not a strong opinion though. If it helps I don't see why not, and user code should not break unless they are using the private api, the public api is everything but the labels module available at the top level.

**Comment by ndmlny-qs** [comment]
Thanks @OriolAbril. Your response about restructuring and your deeper thoughts about the structure in general were what I was interesting in having a conversation about in the "implementation" section above. This conversation will have a lot more meaning when the codebase becomes more complex, and the entry point for future collaborators is higher than at present.

I'm going to assign this issue to me, and structure it have the smallest diff possible, which means that I will just be moving the `ess` evolution calculation out of the plotting functions. A future issue will discuss moving `ess` methods into their own module, and further architectural discussions around the computation methods.

**Comment by OriolAbril** [comment]
> This conversation will have a lot more meaning when the codebase becomes more complex, and the entry point for future collaborators is higher than at present.

Indeed, I think it is clear that there are some "families" of methods (ess, rhat, loo+psis...) but (at least in my case) higher level groupings are not very clear.
